fsw/src/generic_fss_device.c

```diff
@@ -0,0 +1,189 @@
+    {
+        status = GENERIC_FSS_ReadData(device, write_data, read_data, sizeof(read_data));
+        if (status == OS_SUCCESS)
+        {
+            #ifdef GENERIC_FSS_CFG_DEBUG
+                OS_printf("  GENERIC_FSS_RequestData = ");
+                for (uint32_t i = 0; i < sizeof(read_data); i++)
+                {
+                    OS_printf("%02x", read_data[i]);
+                }
+                OS_printf("\n");
+            #endif
+
+            uint8_t checksum = compute_checksum(read_data, 4, 11);
+            /* Verify data header, command code, length, and checksum */
+            if ((read_data[0]  == GENERIC_FSS_DEVICE_HDR_0) && 
+                (read_data[1]  == GENERIC_FSS_DEVICE_HDR_1) && 
+                (read_data[2]  == GENERIC_FSS_DEVICE_HDR_2) && 
+                (read_data[3]  == GENERIC_FSS_DEVICE_HDR_3) &&
+                (read_data[4]  == 0x01)                     &&
+                (read_data[5]  == 0x0A)                     &&
+                (read_data[15] == checksum))
+            {
+                // alpha
+                data->Alpha = fourbytes_little_endian_to_float(&read_data[6]);
+                // beta
+                data->Beta = fourbytes_little_endian_to_float(&read_data[10]);
+                // error code
+                data->ErrorCode = read_data[14];
+                
+                #ifdef GENERIC_FSS_CFG_DEBUG
+                    OS_printf("  Header           = 0x%02x%02x%02x%02x  \n", read_data[0], read_data[1], read_data[2], read_data[3]);
+                    OS_printf("  Command          = 0x%08x              \n", read[4]);
+                    OS_printf("  Length           = 0x%08x              \n", read[5]);
+                    OS_printf("  Checksum         = 0x%08x              \n", read[15]);
+                    OS_printf("  Data alpha       = 0x%04x, %d          \n", data->Alpha, data->Alpha);
+                    OS_printf("  Data beta        = 0x%04x, %d          \n", data->Beta, data->Beta);
+                    OS_printf("  Data error code  = 0x%04x, %d          \n", data->ErrorCode, data->ErrorCode);
+                #endif
+            }
+        } 
+        else
+        {
+            #ifdef GENERIC_FSS_CFG_DEBUG
+                OS_printf("  GENERIC_FSS_RequestData: Invalid data read! \n");
+            #endif 
+            status = OS_ERROR;
+        }
+    }
+    else
+    {
+        #ifdef GENERIC_FSS_CFG_DEBUG
+            OS_printf("  GENERIC_FSS_RequestData: GENERIC_FSS_CommandDevice reported error %d \n", status);
+        #endif 
+    }
+    return status;
+}
+
+/* 
+** Generic read data from device
+*/
+static int32_t GENERIC_FSS_ReadData(spi_info_t *device, uint8_t* write_data, uint8_t* read_data, uint8_t data_length)
+{
+    int32_t status = OS_ERROR;
+    int32_t bytes = 0;
+    int32_t bytes_available = 0;
+    uint8_t ms_timeout_counter = 0;
+
+    /* Read data */
+    if ((spi_select_chip(device) == SPI_SUCCESS) &&
```

Comment:
Is is guaranteed that this statement will be executed sequentially or should select, transaction, unselect be broken apart and checked independently?

Reply:
Yes, due to short circuiting of &&, each condition will only be evaluated if the preceding ones are true.